include/zephyr/rtio/rtio.h

```diff
@@ -0,0 +1,606 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes with low effort
+ *
+ * RTIO uses a SPSC lock-free queue pair to enable a DMA and ISR friendly I/O API.
+ *
+ * I/O like operations are setup in a pre-allocated queue with a fixed number of
+ * submission requests. Each submission request takes the device to operate on
+ * and an operation. The rest is information needed to perform the operation such
+ * as a register or mmio address of the device, a source/destination buffers
+ * to read from or write to, and other pertinent information.
+ *
+ * These operations may be chained in a such a way that only when the current
+ * operation is complete will the next be executed. If the current request fails
+ * all chained requests will also fail.
+ *
+ * The completion of these requests are pushed into a fixed size completion
+ * queue which an application may actively poll for completions.
+ *
+ * An executor (could use a dma controller!) takes the queues and determines
+ * how to perform each requested operation. By default there is a software
+ * executor which does all operations in software using software device
+ * APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_RTIO_RTIO_H_
+#define ZEPHYR_INCLUDE_RTIO_RTIO_H_
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/rtio/rtio_spsc.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief RTIO
+ * @defgroup rtio RTIO
+ * @{
+ * @}
+ */
+
+
+struct rtio_iodev;
+
+/**
+ * @brief RTIO API
+ * @defgroup rtio_api RTIO API
+ * @ingroup rtio
+ * @{
+ */
+
+/**
+ * @brief RTIO Predefined Priorties
+ * @defgroup rtio_sqe_prio RTIO Priorities
+ * @ingroup rtio_api
+ * @{
+ */
+
+/**
+ * @brief Low priority
+ */
+#define RTIO_PRIO_LOW 0
+
+/**
+ * @brief Normal priority
+ */
+#define RTIO_PRIO_NORM 127
+
+/**
+ * @brief High priority
+ */
+#define RTIO_PRIO_HIGH 255
+
+/**
+ * @}
+ */
+
+
+/**
+ * @brief RTIO SQE Flags
+ * @defgroup rtio_sqe_flags RTIO SQE Flags
+ * @ingroup rtio_api
+ * @{
+ */
+
+/**
+ * @brief The next request in the queue should wait on this one.
+ */
+#define RTIO_SQE_CHAINED BIT(0)
+
+/**
+ * @}
+ */
+
+/**
+ * @brief A submission queue event
+ */
+struct rtio_sqe {
+	uint8_t op; /**< Op code */
+
+	uint8_t prio; /**< Op priority */
+
+	uint16_t flags; /**< Op Flags */
+
+	struct rtio_iodev *iodev; /**< Device to operation on */
+
+	/**
+	 * User provided pointer to data which is returned upon operation
+	 * completion
+	 *
+	 * If unique identification of completions is desired this should be
+	 * unique as well.
+	 */
+	void *userdata;
+
+	union {
```

Comment:
Probably more realistically, something like a dmabuf is needed that can understand the dcache/alignment requirements of the platform, at least as one option.

To map to the existing spi/i2c api's it would need to be more like a readv/writev/transcievev (vectored). Maybe that's good enough instead of externalizing those bus transactions.